contracts/v2/core/WithdrawQueue.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {ERC4626, ERC20} from "solmate/src/mixins/ERC4626.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+import {SharedDepositMinterV2} from "./SharedDepositMinterV2.sol";
+
+contract WithdrawQueue is AccessControl, Pausable, ReentrancyGuard {
+    using Address for address payable;
+
+    struct Queue {
+        address to;
+        uint256 amount;
+        address token;
+        uint256 tokenAmount;
+    }
+
+    address public immutable sgEth;
+    address public immutable wsgEth;
+    SharedDepositMinterV2 public immutable minter;
+
+    uint256 public constant CHUNK_SIZE = 32 ether;
+
+    uint256 private front;
+    uint256 private end;
+    uint256 private lockedFront;
+    uint256 private lockedAmount;
+
+    mapping(uint256 => Queue) public queue;
+
+    error NoAvailableQueue();
+
+    constructor(address _sgEth, address _wsgEth, address _minter) {
+        grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        sgEth = _sgEth;
+        wsgEth = _wsgEth;
+        minter = SharedDepositMinterV2(_minter);
+    }
+
+    function push(address to) external payable {
+        _push(to, msg.value, address(0), msg.value);
+    }
+
+    function pushSgEth(address to, uint256 amount) external {
+        ERC20(sgEth).transferFrom(msg.sender, address(this), amount);
+        _push(to, amount, sgEth, amount);
+    }
+
+    function pushWsgEth(address to, uint256 amount) external {
```

Comment:
1. can you prototype out what the redeem phase would look like for this? 
similiar to the 7540 claim fns

2. can we change these names to be more generic? using 7540 as a guide, this would be something like `requestRedeem` for `pushSgETH` and `requestRedeemWrapped` for `pushWsgEth` 

3. push should be internal no? otherwise isnt that a obvious vulnerability? 



@devlancer412

Reply:
erc7540 is a spec that implements what we want here, i.e queues redemptions that can be fulfilled later. 
yea change the names to make it fit, read the spec for logic too. 
overall logic is fine, but missing checks

- if a user can push couldnt they just withdraw all funds in the contract with no collateral?

Reply:
Could you let me know just what contracts should be interacted with this contract?